clients/drcachesim/analyzer.cpp

```diff
@@ -377,50 +441,187 @@ analyzer_tmpl_t<RecordType, ReaderType>::process_tasks(analyzer_worker_data_t *w
+    typename analysis_tool_tmpl_t<RecordType>::interval_state_snapshot_t *two)
+{
+    assert(one != nullptr && two != nullptr);
+    result = tools_[tool_idx]->combine_interval_snapshot(one, two);
+    if (result == nullptr) {
+        error_string_ = "combine_interval_snapshot unexpectedly returned nullptr";
+        return false;
+    }
+    if (!tools_[tool_idx]->release_interval_snapshot(one) ||
+        !tools_[tool_idx]->release_interval_snapshot(two)) {
+        error_string_ = tools_[tool_idx]->get_error_string();
+        return false;
+    }
+    return true;
+}
+
+template <typename RecordType, typename ReaderType>
+bool
+analyzer_tmpl_t<RecordType, ReaderType>::merge_shard_interval_results(
+    // intervals[shard_idx] is a queue of interval_state_snapshot_t*
+    // representing the interval snapshots for that shard.
+    std::vector<std::queue<
+        typename analysis_tool_tmpl_t<RecordType>::interval_state_snapshot_t *>>
+        &intervals,
+    // This function will write the resulting whole-trace intervals to
+    // merged_intervals.
+    std::vector<typename analysis_tool_tmpl_t<RecordType>::interval_state_snapshot_t *>
+        &merged_intervals,
+    int tool_idx)
+{
+    assert(!intervals.empty());
+    assert(merged_intervals.empty());
+    // Used to recompute the interval_id for the result whole trace intervals, which are
+    // numbered by the earliest shard's timestamp.
+    uint64_t earliest_ever_interval_end_timestamp = std::numeric_limits<uint64_t>::max();
+    size_t shard_count = intervals.size();
+    bool any_shard_has_results_left = true;
+    while (any_shard_has_results_left) {
+        // Look for the next whole trace interval across all shards, which will be the
+        // one with the earliest interval-end timestamp.
+        uint64_t earliest_interval_end_timestamp = std::numeric_limits<uint64_t>::max();
+        for (size_t shard_idx = 0; shard_idx < shard_count; ++shard_idx) {
+            if (intervals[shard_idx].empty())
+                continue;
+            earliest_interval_end_timestamp =
+                std::min(earliest_interval_end_timestamp,
+                         intervals[shard_idx].front()->interval_end_timestamp_);
+        }
+        // We're done if no shard has any interval left unprocessed.
+        if (earliest_interval_end_timestamp == std::numeric_limits<uint64_t>::max()) {
+            any_shard_has_results_left = false;
+            continue;
+        }
+        assert(earliest_interval_end_timestamp % interval_microseconds_ == 0);
+        if (earliest_ever_interval_end_timestamp ==
+            std::numeric_limits<uint64_t>::max()) {
+            earliest_ever_interval_end_timestamp = earliest_interval_end_timestamp;
+        }
+
+        // Merge next intervals that have a timestamp == earliest_interval_end_timestamp.
+        // In other words: merge intervals from all shards that were active during this
```

Comment:
@derekbruening 
Re interval results in parallel mode:
We have two possible strategies for merging the shard-local interval snapshots to create the interval snapshots for the whole trace:
1. merge interval snapshots from shards that were active during the interval
2. merge latest interval snapshots for all shards

(1) supports delta counts better. (2) supports cumulative counts better.

This PR implements (1) currently and documents the strategy.

I'm re-considering this though. Since deltas can be computed from cumulative counts during final print_interval_results (where the tool has the vector of all interval's snapshots), maybe it is better that we do (2)?

Reply:
Actually evening cumulative results can be derived from delta results. So maybe it doesn't matter that much?

The other option is that we support both strategies and let the tool decide what should be used by the framework.

Reply:
Changed combine_interval_results to pass all shard's last snapshot, with a per-shard bool telling whether that shard was observed in the last interval. The tool can then choose to use any strategy to combine its fields; can also be field-specific.